tools/perf/builtin-c2c.c

```diff
@@ -2547,9 +3124,804 @@ static void perf_c2c__hists_fprintf(FILE *out, struct perf_session *session)
+		/* Count the number of entries */
+		int symbol_entries = 0;
+		struct rb_node *nd = rb_first_cached(&c2c.symbol_hists.hists.entries);
+		while (nd) {
+			struct hist_entry *he = rb_entry(nd, struct hist_entry, rb_node);
+			if (!he->filtered)
+				symbol_entries++;
+			nd = rb_next(nd);
+		}
+
+		fprintf(out, "\n");
+		fprintf(out, "Shared Data Symbols Table     (%d entries, sorted on Cycles Percent)\n", symbol_entries);
+		fprintf(out, "#\n");
+
+		hists__fprintf(&c2c.symbol_hists.hists, true, 0, 0, 0, stdout, true);
+	}
 }
 
 #ifdef HAVE_SLANG_SUPPORT
+
+/*
+ * Create child hist_entry objects for related symbols
+ * This enables the hierarchical display in the TUI browser
+ * Children are inserted sorted by Stores (stats.store) descending
+ */
+static void populate_symbol_children(struct hist_entry *he)
+{
+	struct c2c_hist_entry *c2c_he;
+    struct related_symbol *rel_sym;
+	struct rb_root_cached *root;
+	struct rb_node **p;
+	struct rb_node *parent = NULL;
+	bool leftmost = true;
+	int count = 0;
+
+	if (!he || !he->has_children)
+		return;
+
+	root = &he->hroot_out;
+
+	/* If already populated, return */
+	if (!RB_EMPTY_ROOT(&root->rb_root))
+		return;
+
+	c2c_he = container_of(he, struct c2c_hist_entry, he);
+	if (!c2c_he)
+		return;
+
+	/* Ensure related_symbols list is valid */
+	if (list_empty(&c2c_he->related_symbols)) {
+		he->has_children = false;  /* Reset inconsistent state */
+		return;
+	}
+
+	/* Create child entries sorted by Stores (stats.store) descending */
+    {
+        int num_rel = 0, idx = 0, i;
+        struct related_symbol **sorted = NULL;
+
+        /* Count related symbols and allocate array in one pass */
+        list_for_each_entry(rel_sym, &c2c_he->related_symbols, list)
+            num_rel++;
+
+        if (num_rel == 0)
+            return;
+
+        sorted = calloc(num_rel, sizeof(*sorted));
+        if (!sorted)
+            return;
+
+        /* Fill array in the same pass */
+        list_for_each_entry(rel_sym, &c2c_he->related_symbols, list)
+            sorted[idx++] = rel_sym;
+
+        /* Simple selection sort by stats.store descending to avoid qsort dependency nuances */
+        for (i = 0; i < num_rel - 1; i++) {
+            int max_j = i, j;
+            for (j = i + 1; j < num_rel; j++) {
+                uint64_t a = (uint64_t)sorted[max_j]->stats.store;
+                uint64_t b = (uint64_t)sorted[j]->stats.store;
+                if (b > a)
+                    max_j = j;
+            }
+            if (max_j != i) {
+                struct related_symbol *tmp = sorted[i];
+                sorted[i] = sorted[max_j];
+                sorted[max_j] = tmp;
+            }
+        }
+
+        /* Emit children in sorted order */
+        for (i = 0; i < num_rel; i++) {
+            struct c2c_hist_entry *child_c2c_he;
+            struct hist_entry *child_he;
+            rel_sym = sorted[i];
+
+		if (!rel_sym || !rel_sym->sym) {
+			continue;
+        }
```

Comment:
seems there is indent mismatch, try to format with clang-format:
https://www.kernel.org/doc/html/next/dev-tools/clang-format.html